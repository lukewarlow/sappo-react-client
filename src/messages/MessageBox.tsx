import React from "react";
import ClickNHold from 'react-click-n-hold';
import {isSpeciesMessage, Message} from "../api/MessagesApi";
import {MapModal} from "./MapModal";
import styles from "./MessageBox.module.css";
import {formatDistanceStrict} from "date-fns";

export default class MessageBox extends React.Component<IMessageBoxProps, IMessageBoxState> {
    state: IMessageBoxState = {
        showMapModal: false
    };

    render() {
        let cardContents: JSX.Element;
        let relativeTime = formatDistanceStrict(new Date(this.props.message.timestamp), new Date(Date.now()));

        if (isSpeciesMessage(this.props.message)) {
            cardContents = (
                <ClickNHold
                    time={0.7}
                    onClickNHold={(e) => {this.setState({showMapModal: true})}}
                    onStart={_ => {}}
                    onEnd={_ => {}}>
                    <div className="card-body" title={`${relativeTime} ago`}>
                        <div><span className="font-weight-bold">{this.props.message.username}</span>: {this.props.message.content}</div>
                        <div>Species: {this.props.message.species}</div>
                        <div>Abundance: {this.props.message.abundance}</div>
                        <div>Temperature: {this.props.message.temperature}Â°C</div>
                        {this.props.message.image && <img className="img-fluid" alt={`${this.props.message.username}'s image`} src={this.props.message.image} />}
                    </div>
                </ClickNHold>
            )
        } else {
            cardContents = (
                <div className="card-body" title={`${relativeTime} ago`}>
                    <div className="font-weight-bold">{this.props.message.username}: {this.props.message.content}</div>
                </div>
            )
        }

        return (
            <div className="d-inline-block w-100 mb-2">
                <div className={`card ${isSpeciesMessage(this.props.message) ? 'border-success' : ''} ${this.props.message.username == this.props.currentUsername ? 'bg-success float-right' : 'bg-light'} ${styles['message-box']}`}>
                    {cardContents}
                </div>
                {this.state.showMapModal && isSpeciesMessage(this.props.message) && <MapModal onClose={() => this.setState({showMapModal: false})} message={this.props.message} />}
            </div>
        )
    }
}

export interface IMessageBoxProps {
    readonly message: Message,
    readonly currentUsername: string
}

export interface IMessageBoxState {
    showMapModal: boolean
}
