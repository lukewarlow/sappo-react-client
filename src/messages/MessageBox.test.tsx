import React from 'react';
import ReactDOM from 'react-dom';
import MessageBox from "./MessageBox";
import renderer from "react-test-renderer";
import {emptyMessage} from "../api/MessagesApi";

it('MessageBox renders without crashing', () => {
  const div = document.createElement('div');
  const username = "Test User";
  ReactDOM.render(<MessageBox message={emptyMessage(username)} currentUsername={username}/>, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('MessageBox renders correctly', () => {
  const username = "Test User";
  let realDateNow = Date.now.bind(global.Date);
  let dateNowStub = jest.fn(() => 1573080910769);
  global.Date.now = dateNowStub;
  let message = emptyMessage(username);
  message.timestamp = new Date("2019-11-01");
  message.content = "Hello Everyone";
  const tree = renderer
      .create(<MessageBox message={message} currentUsername={username}/>)
      .toJSON();

  expect(tree).toMatchSnapshot();
  expect(dateNowStub).toHaveBeenCalled();
  global.Date.now = realDateNow;
});
