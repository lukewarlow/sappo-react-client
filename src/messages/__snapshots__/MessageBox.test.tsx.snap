// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageBox renders correctly 1`] = `
<div
  className="d-inline-block w-100 mb-2"
>
  <div
    className="card  bg-success float-right message-box"
  >
    <div
      className="card-body"
      title="6 days ago"
    >
      <div
        className="font-weight-bold"
      >
        Test User
        : 
        Hello Everyone
      </div>
    </div>
  </div>
</div>
`;
